spec/datadog/profiling/collectors/discrete_dynamic_sampler_spec.rb

```diff
@@ -0,0 +1,226 @@
+        # up and eventually we can sample again within some reasonable amount of settings (15 secs
+        last_p = p1
+        saw_sample = 10.times do
+          stats = simulate_load(duration_seconds: 2, events_per_second: 1, sampling_seconds: 0.1)
+          current_p = sampler._native_probability
+          expect(current_p).to be > last_p
+
+          last_p = current_p
+
+          break true if stats[:num_samples] > 0
+        end
+        expect(saw_sample).to be(true)
+      end
+    end
+  end
+
+  context 'when sampling time worsens' do
+    it 'will readjust to decrease sampling rate' do
+      # Start with an initial load of 8 eps @ 0.01s sampling time should give us a sampling
+      # probability of around 25% given our 2% overhead target (see similar test case above)
+      stats = simulate_load(duration_seconds: 60, events_per_second: 8, sampling_seconds: 0.01)
+      expect(stats[:sampling_ratio]).to be_between(0.23, 0.27)
+
+      # However, we'll now worsen our sampling time in 2x keeping all other things equal. We
+      # expect our sampling probability to halve (~12.5%)
+      stats = simulate_load(duration_seconds: 60, events_per_second: 8, sampling_seconds: 0.02)
+      expect(stats[:sampling_ratio]).to be_between(0.10, 0.13)
+    end
+  end
+
+  context 'when sampling time improves' do
+    it 'will readjust to increase sampling rate' do
+      # Start with an initial load of 8 eps @ 0.01s sampling time should give us a sampling
+      # probability of around 25% given our 2% overhead target (see similar test case above)
+      stats = simulate_load(duration_seconds: 60, events_per_second: 8, sampling_seconds: 0.01)
+      expect(stats[:sampling_ratio]).to be_between(0.23, 0.27)
+
+      # However, we'll now improve our sampling time in 2x keeping all other things equal. We
+      # expect our sampling probability to double (~50%)
+      stats = simulate_load(duration_seconds: 60, events_per_second: 8, sampling_seconds: 0.005)
+      expect(stats[:sampling_ratio]).to be_between(0.45, 0.55)
+    end
+  end
+
+  context 'given a constant load' do
+    it "the higher the target overhead, the more we'll sample" do
+      # Warm-up to overcome initial hardcoded window
+      simulate_load(duration_seconds: 5, events_per_second: 4, sampling_seconds: 0.01)
+      # Start with an initial load of 4 eps @ 0.01s sampling time should give us a sampling
+      # probability of around 50% given our 2% overhead target (see similar test case above)
+      stats = simulate_load(duration_seconds: 60, events_per_second: 4, sampling_seconds: 0.01)
+      expect(stats[:sampling_ratio]).to be_between(0.45, 0.55)
+
+      # We'll now increase our overhead target to 4%
+      update_overhead_target(max_overhead_target * 2)
+
+      # Warm-up to overcome initial hardcoded window
+      simulate_load(duration_seconds: 5, events_per_second: 4, sampling_seconds: 0.01)
```

Comment:
Minor: I wonder if we should instead introduce a helper to set the current state (or initial state) in a more white-box way.

Reply:
Hmm mixed feelings about it. I guess it makes it easier to reason about the starting state but we risk setting up tests with parameters that would never occur in practice and may hide issues?

Reply:
Yeah, it's always the big downside of such things :/ Well, if it gets too annoying it may be an option, but I think for now we can maybe skip it.